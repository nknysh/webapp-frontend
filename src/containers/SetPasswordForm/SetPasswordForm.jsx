import React, { useState } from 'react';
import { Redirect } from 'react-router-dom';
import { __, compose, prop, path, view, set, curry, keys } from 'ramda';

import { Form, Label, Loader, Title, Fields } from 'components';

import { Status } from 'store/common';
import { InputError } from 'styles/elements';
import uiConfig from 'config/ui';
import formConfig from 'config/forms';
import { schema, fields } from 'config/forms/setPassword';
import { lensesFromObject } from 'utils';

import connect from './SetPasswordForm.state';
import { propTypes, defaultProps } from './SetPasswordForm.props';
import { StyledSetPasswordForm, Field, Input, Actions, SubmitButton, SubmitText } from './SetPasswordForm.styles';

const renderFormError = (key, errors) => prop(key, errors) && <InputError>{prop(key, errors)}</InputError>;

export const SetPasswordForm = ({ requestStatus, onSetPassword, token }) => {
  // No token, no form
  if (!token) return <Redirect to="/" />;

  const [submitted, setSubmitted] = useState(false);
  const [formValues, setFormValues] = useState(prop('defaults', fields));

  const isSending = requestStatus === Status.SENDING;
  const success = requestStatus === Status.SUCCESS;

  const lenses = lensesFromObject(keys(formValues));
  const getLens = prop(__, lenses);

  const changeHandler = curry((lens, callback, e) => {
    setFormValues(set(lens, e.target.value, formValues));
    callback(e);
  });

  const onSubmit = values => {
    setSubmitted(true);
    setFormValues(values);
    onSetPassword({ values, token });
  };

  const renderForm = () => (
    <Form
      initialValues={formValues}
      onSubmit={onSubmit}
      validationSchema={schema}
      validateOnBlur={false}
      validateOnChange={false}
    >
      {({ errors, handleChange, handleBlur, handleSubmit }) => (
        <form>
          <Fields>
            <Field>
              <Label htmlFor="password">{path(['labels', 'password'], fields)}</Label>
              <Input
                autocomplete
                type="password"
                name="password"
                placeholder={path(['labels', 'password'], fields)}
                value={view(getLens('password'), formValues)}
                onChange={changeHandler(getLens('password'), handleChange)}
                onBlur={handleBlur}
              />
              {renderFormError('password', errors)}
            </Field>
            <Field>
              <Label htmlFor="confirm">{path(['labels', 'confirm'], fields)}</Label>
              <Input
                autocomplete
                type="password"
                name="confirm"
                placeholder={path(['labels', 'confirm'], fields)}
                value={view(getLens('confirm'), formValues)}
                onChange={changeHandler(getLens('confirm'), handleChange)}
                onBlur={handleBlur}
              />
              {renderFormError('confirm', errors)}
            </Field>
          </Fields>
          <Actions>
            <SubmitButton type="button" onClick={handleSubmit}>
              <SubmitText>{path(['buttons', 'submit'], uiConfig)}</SubmitText>
            </SubmitButton>
          </Actions>
        </form>
      )}
    </Form>
  );

  const isComplete = !isSending && success;

  const title = !isComplete
    ? path(['forms', 'setPassword'], formConfig)
    : path(['forms', 'setPasswordComplete'], formConfig);

  return (
    <StyledSetPasswordForm>
      <Loader isLoading={submitted && isSending && !success} text={path(['messages', 'loggingIn'], uiConfig)}>
        <Title>{title}</Title>
        {!isComplete && renderForm()}
      </Loader>
    </StyledSetPasswordForm>
  );
};

SetPasswordForm.propTypes = propTypes;
SetPasswordForm.defaultProps = defaultProps;

export default compose(connect)(SetPasswordForm);
