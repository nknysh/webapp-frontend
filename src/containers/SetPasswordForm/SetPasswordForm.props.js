import PropTypes from 'prop-types';

import { noop } from 'utils';

export const propTypes = {
  onReset: PropTypes.func,
  requestStatus: PropTypes.string,
};

export const defaultProps = {
  onReset: noop,
};
