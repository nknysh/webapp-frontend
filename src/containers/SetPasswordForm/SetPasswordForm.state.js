import { connect } from 'react-redux';
import { pipe } from 'ramda';

import { setPassword } from 'store/modules/auth/actions';
import { getAuthStatus, getAuthError } from 'store/modules/auth/selectors';

export const mapStateToProps = state => ({
  requestStatus: getAuthStatus(state),
  error: getAuthError(state),
});

export const mapDispatchToProps = dispatch => ({
  onSetPassword: pipe(
    setPassword,
    dispatch
  ),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
);
