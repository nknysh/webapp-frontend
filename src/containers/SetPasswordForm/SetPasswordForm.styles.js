import styled from 'styled-components';

import { Checkbox, Markdown } from 'components';

import theme from 'styles/theme';
import breakpoints from 'styles/breakpoints';
import { ServerError } from 'styles/errors';
import { Input as BaseInput, Button as BaseButton } from 'styles/elements';

export const StyledSetPasswordForm = styled.div`
  padding: ${theme.gutter}px;

  ${breakpoints.tablet`
    padding: ${theme.gutter * 5}px;
  `}
`;

export const Fields = styled.div`
  margin-top: 50px;
  text-align: left;

  ${breakpoints.tablet`
    min-width: 400px;
  `}
`;

export const Field = styled.div`
  margin-bottom: ${theme.gutter * 2}px;
`;

export const Input = styled(BaseInput)``;

export const Actions = styled.div`
  margin-top: 50px;
`;

export const SubmitButton = styled(BaseButton)``;

export const SubmitText = styled.span`
  color: ${theme.colors.white};
`;

export const StyledCheckbox = styled(Checkbox)`
  label span {
    text-transform: uppercase !important;
  }
`;

export const ServerErrorContent = styled(ServerError)``;

export const StyledMarkdown = styled(Markdown)`
  text-align: center;

  p {
    margin: ${theme.gutter / 2}px 0;
  }

  ${breakpoints.tablet`
    width: 400px;
  `}
`;
