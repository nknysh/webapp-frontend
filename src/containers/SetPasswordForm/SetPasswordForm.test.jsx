import React from 'react';

import { SetPasswordForm } from './SetPasswordForm';

const getComponent = props => shallow(<SetPasswordForm {...props} />);

describe('<SetPasswordForm />', () => {
  describe('render', () => {
    it('matches snapshot', () => {
      expect(getComponent()).toMatchSnapshot();
    });
  });
});
