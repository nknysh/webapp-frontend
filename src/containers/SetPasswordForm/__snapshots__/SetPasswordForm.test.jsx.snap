// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SetPasswordForm /> render matches snapshot 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;
